### ISSUE TITLE
needless_doctest_main yields false positive with proc macros

### ISSUE THREAD

**Issue by Rantanen** [rust-lang/rust-clippy#4698]
Clippy version: 0.0.212 (e979eb4c 2019-10-17)

Clippy warns that the `fn main() {}` is not needed in the following doctest:

https://github.com/Rantanen/intercom/blob/45837a4d77c340c9cb8cf45142843198a80a82b6/intercom/src/lib.rs#L12-L36

However removing that will end up with doctest wrapping the whole thing in a function, which results in the `com_library!(Calculator);` becoming a statement - something that is not allowed for (at least that) proc macro.

I feel like this is enough of a niche case that there's no real need to fix it. A note could be added to the Known problems section of the lint.

**Comment by Rantanen** [comment]
Okay this was a bit more annoying than I thought. The fact that I had the issue on crate level documentation prevented me from using `#[allow(clippy::needless_doctest_main)]` on that specific doctest. The attribute _inside_ the doctest did nothing while outside the only way I could have put it on the "item" was a crate-wide `#![allow(...)]`.

In the end I found out that the check is a rather simple one:

https://github.com/rust-lang/rust-clippy/blob/850ad526df037f8eb1e2e4a596f08c5e3efab76e/clippy_lints/src/doc.rs#L347-L351

My workaround involves turning the main function into `fn main() -> () {}`, which results in Clippy not discovering it.

**Comment by llogiq** [comment]
Yeah the check is at the moment just text-based. And detecting proc macros that expand to items is somewhat harder.

A simple trick that would work is
```
///# fn main 
///# // keep this to quell `needless_doctest_main` warning
///#(){}
```
